Compiler warning when building without upstream app_drivers enabled

### Operating System

Linux

### Commit SHA

since 652342b57b129a79e0df152377d18096e4a0fed5

### Board

SAMD21

### Firmware

https://github.com/DO3RB/WirelessNetworkTransceiver/tree/master/code_rfm12bp
compiled on Linux 7.1.2-arch with arm-none-eabi-gcc 16.1.0 and arm-none-eabi-newlib 4.6.0.20260123

### What happened ?

Dear Team, thank you for your work.

I use TinyUSB in an admittedly unusual configuration: Without builtin app_drivers enabled. This produces a warning for quite some time now, which bothers me.

Please be so kind and consider my following summary, and proposed changes :)

---

Function `get_driver` in [src/device/usbd.c](https://github.com/hathach/tinyusb/blob/0.19.0/src/device/usbd.c#L340) (release 0.19)
---

``` c
enum { BUILTIN_DRIVER_COUNT = TU_ARRAY_SIZE(_usbd_driver) }; // TRICK: compile-time constant <<<

// Additional class drivers implemented by application
tu_static usbd_class_driver_t const * _app_driver = NULL;
tu_static uint8_t _app_driver_count = 0;

#define TOTAL_DRIVER_COUNT    ((uint8_t) (_app_driver_count + BUILTIN_DRIVER_COUNT))

// virtually joins built-in and application drivers together.
// Application is positioned first to allow overwriting built-in ones.
TU_ATTR_ALWAYS_INLINE static inline usbd_class_driver_t const * get_driver(uint8_t drvid) {
  usbd_class_driver_t const *driver = NULL;
  if (drvid < _app_driver_count) {
    // Application drivers
    driver = &_app_driver[drvid];
  } else if (drvid < TOTAL_DRIVER_COUNT && BUILTIN_DRIVER_COUNT > 0) { // CHECK if builtins enabled <<<
    driver = &_usbd_driver[drvid - _app_driver_count];
  }
  return driver;
}
```

**produced no GCC warnings.**


Since [fix more alerts found by pvs-studio](https://github.com/hathach/tinyusb/commit/652342b57b129a79e0df152377d18096e4a0fed5#diff-547ff7242684ae3243ee7fa7a3fb7a6efc3f2d2bae3e4ba4908b281f47b871d4) (release 0.20)
---

``` c
enum { BUILTIN_DRIVER_COUNT = TU_ARRAY_SIZE(_usbd_driver) }; // TRICK persists <<<

// Additional class drivers implemented by application
static const usbd_class_driver_t *_app_driver       = NULL;
static uint8_t                    _app_driver_count = 0;

#define TOTAL_DRIVER_COUNT    ((uint8_t) (_app_driver_count + BUILTIN_DRIVER_COUNT))

// virtually joins built-in and application drivers together.
// Application is positioned first to allow overwriting built-in ones.
TU_ATTR_ALWAYS_INLINE static inline usbd_class_driver_t const * get_driver(uint8_t drvid) {
  usbd_class_driver_t const *driver = NULL;
  if (drvid < _app_driver_count) {
    // Application drivers
    driver = &_app_driver[drvid];
  } else{
    drvid -= _app_driver_count;
    if (drvid < BUILTIN_DRIVER_COUNT) { // CHECK is gone <<<
      driver = &_usbd_driver[drvid];
    }
  }
  return driver;
}
```

**produces GCC warning** (which i have to patch every subsequent release):

``` sh
$ make
arm-none-eabi-gcc -Wall -Wextra -O2 -flto -ffunction-sections -fdata-sections -fwrapv -fdiagnostics-color -I tinyusb -I. -MMD -MP -I tinyusb -DCFG_TUSB_CONFIG_FILE=\"code_rfm12bp/tinyusb.h\" -mcpu=cortex-m0plus -mthumb -c tinyusb/device/usbd.c -o code_rfm12bp/.build/tinyusb/device/usbd.o
tinyusb/device/usbd.c: In function 'get_driver':
tinyusb/device/usbd.c:400:43: warning: comparison is always false due to limited range of data type [-Wtype-limits]
  400 |     if (drvid < BUILTIN_DRIVER_COUNT) {
      |               ^
arm-none-eabi-gcc -Wall -Wextra -O2 -flto -ffunction-sections -fdata-sections -fwrapv -fdiagnostics-color -I code_rfm12bp -I. -MMD -MP -I tinyusb -DCFG_TUSB_CONFIG_FILE=\"code_rfm12bp/tinyusb.h\" -nostdlib -T samd21g18a/sam.ld -Wl,--gc-sections -mcpu=cortex-m0plus -mthumb code_rfm12bp/.build/code_rfm12bp/main.o code_rfm12bp/.build/samd21g18a/syscalls.o code_rfm12bp/.build/samd21g18a/newlib.o code_rfm12bp/.build/code_rfm12bp/tinyusb.o code_rfm12bp/.build/samd21g18a/fiber.o code_rfm12bp/.build/samd21g18a/sam.o code_rfm12bp/.build/tinyusb/device/usbd.o code_rfm12bp/.build/code_rfm12bp/storage.o code_rfm12bp/.build/code_rfm12bp/circuit.o code_rfm12bp/.build/tinyusb/common/tusb_fifo.o code_rfm12bp/.build/code_rfm12bp/rfm12bp.o code_rfm12bp/.build/code_rfm12bp/conduit.o code_rfm12bp/.build/tinyusb/tusb.o code_rfm12bp/.build/samd21g18a/console.o code_rfm12bp/.build/code_rfm12bp/packets.o code_rfm12bp/.build/code_rfm12bp/network.o code_rfm12bp/.build/code_rfm12bp/channel.o code_rfm12bp/.build/samd21g18a/tinyusb_acm.o code_rfm12bp/.build/samd21g18a/tinyusb_ecm.o code_rfm12bp/.build/code_rfm12bp/do3rb.o code_rfm12bp/.build/tinyusb/device/dcd.o code_rfm12bp/.build/samd21g18a/version.o code_rfm12bp/.build/code_rfm12bp/popcount.o -Wl,--start-group -lm -lc_nano -lgcc -Wl,--end-group -o code_rfm12bp/.build/main.elf
arm-none-eabi-size
section     size addr
.boot       8192 0x00000000
.text      37804 0x00002000 # program size for enum-trick comparison
.store       256 0x0003F000
.bss        1360 0x20000000
.noinit    10248 0x20000550
.data        473 0x20002D58
.heap      11468 0x20002F34
.stack      9216 0x20005C00
```


Proposed changes to [src/device/usbd.c](https://github.com/hathach/tinyusb/blob/0.21.0/src/device/usbd.c#L383) (release 0.21)
----

``` c
  static const uint8_t BUILTIN_DRIVER_COUNT = TU_ARRAY_SIZE(_usbd_driver); // CHANGE: enum to const <<<
     
  // Additional class drivers implemented by application   
  static const usbd_class_driver_t *_app_driver       = NULL;   
  static uint8_t                    _app_driver_count = 0;   
     
  #define TOTAL_DRIVER_COUNT    ((uint8_t) (_app_driver_count + BUILTIN_DRIVER_COUNT))   
     
  // virtually joins built-in and application drivers together.   
  // Application is positioned first to allow overwriting built-in ones.   
  TU_ATTR_ALWAYS_INLINE static inline usbd_class_driver_t const * get_driver(uint8_t drvid) {   
    usbd_class_driver_t const *driver = NULL;   
    if (drvid < _app_driver_count) {   
      // Application drivers   
      driver = &_app_driver[drvid];
    } else {
      drvid -= _app_driver_count;
      if (BUILTIN_DRIVER_COUNT > 0 && drvid < BUILTIN_DRIVER_COUNT) { // CHANGE: restore check count > 0 in front, so that it short-circuits the second test <<<
        driver = &_usbd_driver[drvid];
      }
    }
    return driver;
  }
```

**produce no GCC warning, and add no bloat.**

``` sh
$ make
arm-none-eabi-gcc -Wall -Wextra -O2 -flto -ffunction-sections -fdata-sections -fwrapv -fdiagnostics-color -I tinyusb -I. -MMD -MP -I tinyusb -DCFG_TUSB_CONFIG_FILE=\"code_rfm12bp/tinyusb.h\" -mcpu=cortex-m0plus -mthumb -c tinyusb/device/usbd.c -o code_rfm12bp/.build/tinyusb/device/usbd.o
# warning appeased :)
arm-none-eabi-gcc -Wall -Wextra -O2 -flto -ffunction-sections -fdata-sections -fwrapv -fdiagnostics-color -I code_rfm12bp -I. -MMD -MP -I tinyusb -DCFG_TUSB_CONFIG_FILE=\"code_rfm12bp/tinyusb.h\" -nostdlib -T samd21g18a/sam.ld -Wl,--gc-sections -mcpu=cortex-m0plus -mthumb code_rfm12bp/.build/code_rfm12bp/main.o code_rfm12bp/.build/samd21g18a/syscalls.o code_rfm12bp/.build/samd21g18a/newlib.o code_rfm12bp/.build/code_rfm12bp/tinyusb.o code_rfm12bp/.build/samd21g18a/fiber.o code_rfm12bp/.build/samd21g18a/sam.o code_rfm12bp/.build/tinyusb/device/usbd.o code_rfm12bp/.build/code_rfm12bp/storage.o code_rfm12bp/.build/code_rfm12bp/circuit.o code_rfm12bp/.build/tinyusb/common/tusb_fifo.o code_rfm12bp/.build/code_rfm12bp/rfm12bp.o code_rfm12bp/.build/code_rfm12bp/conduit.o code_rfm12bp/.build/tinyusb/tusb.o code_rfm12bp/.build/samd21g18a/console.o code_rfm12bp/.build/code_rfm12bp/packets.o code_rfm12bp/.build/code_rfm12bp/network.o code_rfm12bp/.build/code_rfm12bp/channel.o code_rfm12bp/.build/samd21g18a/tinyusb_acm.o code_rfm12bp/.build/samd21g18a/tinyusb_ecm.o code_rfm12bp/.build/code_rfm12bp/do3rb.o code_rfm12bp/.build/tinyusb/device/dcd.o code_rfm12bp/.build/samd21g18a/version.o code_rfm12bp/.build/code_rfm12bp/popcount.o -Wl,--start-group -lm -lc_nano -lgcc -Wl,--end-group -o code_rfm12bp/.build/main.elf
arm-none-eabi-size
section     size addr
.boot       8192 0x00000000
.text      37804 0x00002000 # program size unchanged :D
.store       256 0x0003F000
.bss        1360 0x20000000
.noinit    10248 0x20000550
.data        473 0x20002D58
.heap      11468 0x20002F34
.stack      9216 0x20005C00
```

### How to reproduce ?

.

### Debug Log as txt file (LOG/CFG_TUSB_DEBUG=2)

.

### Screenshots

_No response_

### I have checked existing issues, discussion and documentation

- [x] I confirm I have checked existing issues, discussion and documentation.

--- a/src/device/usbd.c
+++ b/src/device/usbd.c
@@ -361,13 +361,12 @@ static const usbd_class_driver_t _usbd_driver[] = {
     #endif
 };
 
-enum { BUILTIN_DRIVER_COUNT = TU_ARRAY_SIZE(_usbd_driver) };
-
 // Additional class drivers implemented by application
-static const usbd_class_driver_t *_app_driver       = NULL;
-static uint8_t                    _app_driver_count = 0;
+static const usbd_class_driver_t *_app_driver = NULL;
+static const uint8_t _builtin_driver_count    = TU_ARRAY_SIZE(_usbd_driver);
+static uint8_t              _app_driver_count = 0;
 
-#define TOTAL_DRIVER_COUNT    ((uint8_t) (_app_driver_count + BUILTIN_DRIVER_COUNT))
+#define TOTAL_DRIVER_COUNT    ((uint8_t) (_app_driver_count + _builtin_driver_count))
 
 // virtually joins built-in and application drivers together.
 // Application is positioned first to allow overwriting built-in ones.
@@ -378,7 +377,7 @@ TU_ATTR_ALWAYS_INLINE static inline usbd_class_driver_t const * get_driver(uint8
     driver = &_app_driver[drvid];
   } else{
     drvid -= _app_driver_count;
-    if (BUILTIN_DRIVER_COUNT > 0 && drvid < BUILTIN_DRIVER_COUNT) {
+    if (_builtin_driver_count > 0 && drvid < _builtin_driver_count) {
       driver = &_usbd_driver[drvid];
     }
   }
@@ -563,7 +562,7 @@ bool tud_rhport_init(uint8_t rhport, const tusb_rhport_init_t* rh_init) {
 
   // Get application driver if available
   _app_driver = usbd_app_driver_get_cb(&_app_driver_count);
-  TU_ASSERT(_app_driver_count + BUILTIN_DRIVER_COUNT <= UINT8_MAX);
+  TU_ASSERT(_app_driver_count + _builtin_driver_count <= UINT8_MAX);
 
   // Init class drivers
   for (uint8_t i = 0; i < TOTAL_DRIVER_COUNT; i++) {

--- a/src/device/usbd.c
+++ b/src/device/usbd.c
@@ -378,7 +378,7 @@ TU_ATTR_ALWAYS_INLINE static inline usbd_class_driver_t const * get_driver(uint8
     driver = &_app_driver[drvid];
   } else{
     drvid -= _app_driver_count;
-    if (drvid < BUILTIN_DRIVER_COUNT) {
+    if (BUILTIN_DRIVER_COUNT > 0 && drvid < BUILTIN_DRIVER_COUNT) {
       driver = &_usbd_driver[drvid];
     }
   }

--- a/src/host/usbh.c
+++ b/src/host/usbh.c
@@ -319,13 +319,12 @@ static usbh_class_driver_t const usbh_class_drivers[] = {
   #endif
 };
 
-enum { BUILTIN_DRIVER_COUNT = TU_ARRAY_SIZE(usbh_class_drivers) };
-
 // Additional class drivers implemented by application
 static usbh_class_driver_t const * _app_driver = NULL;
-static uint8_t _app_driver_count = 0;
+static const uint8_t _builtin_driver_count     = TU_ARRAY_SIZE(usbh_class_drivers);
+static uint8_t _app_driver_count               = 0;
 
-#define TOTAL_DRIVER_COUNT    (_app_driver_count + BUILTIN_DRIVER_COUNT)
+#define TOTAL_DRIVER_COUNT    (_app_driver_count + _builtin_driver_count)
 
 // virtually joins built-in and application drivers together.
 // Application is positioned first to allow overwriting built-in ones.
@@ -335,7 +334,7 @@ TU_ATTR_ALWAYS_INLINE static inline usbh_class_driver_t const *get_driver(uint8_
     driver = &_app_driver[drv_id];
   } else {
     drv_id -= _app_driver_count;
-    if (drv_id < BUILTIN_DRIVER_COUNT) {
+    if (_builtin_driver_count > 0 && drv_id < _builtin_driver_count) {
       driver = &usbh_class_drivers[drv_id];
     }
   }
@@ -547,6 +546,7 @@ bool tuh_rhport_init(uint8_t rhport, const tusb_rhport_init_t* rh_init) {
 
     // Get application driver if available
     _app_driver = usbh_app_driver_get_cb(&_app_driver_count);
+    TU_ASSERT(_app_driver_count + _builtin_driver_count <= UINT8_MAX);
 
     // Device
     tu_memclr(_usbh_devices, sizeof(_usbh_devices));

